_drafts/profiling-firmware-on-cortex-m.md

```diff
@@ -0,0 +1,682 @@
+frames to the LCD over SPI.
+
+Unfortunately, we cannot get much more precise results using this approach.
+Using GDB and OpenOCD introduces huge overhead, which means that we are limited
+to a datapoint every 50ms. Given that our MPU executes on the order of 200
+million clock cycles per second, that means that we get a sample every 10
+million clock cycle. This is too coarse for any meaningful analysis, so we must
+find a better approach
+
+## PC Sampling with ITM
+
+To get better profiling data, we turn to the Instrumented Trace Macrocell (ITM).
+The ITM is an optional feature of ARM Cortex-M cores which formats and outputs
+trace information generated by the firmware or directly from the hardware over a
+dedicated bus.
+
+Typically ITM is used to print out debug data from firmware, much like one would
+use `printf` over a serial bus. ITM is strictly superior to UART in this case as
+it is faster, work directly with your SWD adapter, and is buffered so the CPU
+does not need to wait. You can read more about using ITM for logging in Jorge
+Aparicio's [excellent post on the topic](https://blog.japaric.io/itm/).
+
+Where ITM really shines is its ability to output data generated directly by the
+hardware. As of this writing, ITM supports the following hardware data sources:
+
+1. Event counter wrapping: When enabled, an event is generated when DWT
+   event counters overflow. This includes clock cycle counters, folded
+instruction  counters, sleep cycles counters, and more[^dwt].
+2. Exception tracing: When enabled, an event is generated whenever an exception
+   (i.e. Interrupt) occurs.
+3. PC sampling: When enabled, an event is generated at a regular interval
+   containing the current PC.
+4. DWT Data trace: When enabled, events are generated when a specific data
+   address (or range of addresses) is accessed. The events contain the PC, the
+data address, and the data value.
+
+This information can be read over the TRACE pins of your MCU if you have a
+trace-enabled debugger (e.g. a J-Trace or a Lauterbach TRACE32), alternatively
+it can be streamed over the SWO pin of your SWD bus. While more pin allow for
+more data to be read, much can be done with the single SWO pin without requiring
+complex tools.
+
+> Note: Using ITM for printf is relatively straightforward. Here's a putc
+> implementation that should do the trick:
+>
+>     void putc(char c) {
+>       while (ITM_STIM32(0) == 0);
+>       ITM_STIM8(0) = c;
+>     }
+>
+> You can then just `cat` the `itm.fifo` file we configure with openOCD below
+> and see the print output. It's pretty cool!
+
+### Enabling PC Sampling with ITM and OpenOCD
+
+You will notice that "PC Sampling" is explicitly called out as one of the things
+we can do with ITM. With ITM PC Sampling, we can build a much better Sampling
+Profiler!
+
+#### Configuring OpenOCD
+
+First, we must configure OpenOCD to read from the SWO pin and dump the data
```

Comment:
Maybe tell the reader what the TPIU actually is, and then use the abbreviation. 

The common way to do this is to configure the Trace Port Interface Unit, or the TPIU[^tpiu].

[^tpiu]: [Trace Port Interface Unit](https://developer.arm.com/docs/100166/latest/trace-port-interface-unit/about-the-tpiu)

Reply:
:+1: